src/type_checker/mod.rs

```diff
@@ -126,7 +127,17 @@ impl<B: Backend> TypeChecker<B> {
                 sum
             }
             TyKind::BigInt => 1,
-            TyKind::Array(typ, len) => (*len as usize) * self.size_of(typ),
+            TyKind::Array(typ, size) => match size {
+                ArraySize::Number(n) => (*n as usize) * self.size_of(typ),
+                ArraySize::ConstVar(var) => {
```

Comment:
at this point I'm confused, I would have expected sizeof to either return a symbolic value, or an actual value. But here we manage to actually return an actual value. This should only work when we use something like:

```rust
const val = 4;
some_call([0; val]);
```

right? but not with const generics

Reply:
That is a nice catch. The `size_of` in type checker should only return symbolic value. Then in the circuit generation it should manage to return an actual value.

Reply:
only return symbolic value IF there's a const generic involved

Reply:
That makes sense. The type checker shouldn't tracing the actual value for the const vars.

So we may need to change the function return
```rust
fn size_of(&self, typ: &TyKind) -> usize
```

to 

```rust
fn size_of(&self, typ: &TyKind) -> SymbolicValue
```

where `SymbolicValue` can represent something like:

```
cst * 10 + 5
```

Reply:
yeah that makes sense, something like:

```rust
enum Symbolic {
  Add(Symbolic, Symbolic),
  Mul(Symbolic, Symbolic),
  Concrete(BigInt)
}
```